include/deal.II/grid/manifold_lib.h

```diff
@@ -503,6 +504,282 @@ class TorusManifold : public ChartManifold<dim,3,3>
+   */
+  std::array<unsigned int, 10>
+  get_possible_cells_around_points(const std::vector<Point<spacedim> > &surrounding_points) const;
+
+  /**
+   * Finalizes the identification of the correct chart and returns the location
+   * of the surrounding points on the chart. This method internally calls
+   * @p get_possible_cells_around_points
+   */
+  std::pair<typename Triangulation<dim,spacedim>::cell_iterator,
+      std::vector<Point<dim> > >
+      compute_chart_points(const std::vector<Point<spacedim> > &surrounding_points) const;
+
+  /**
+   * Pull back operation into the unit coordinates on the given coarse cell.
+   *
+   * @note This internal function is currently not compatible with the
+   * ChartManifold::pull_back() function because the given class represents an
+   * atlas of charts, not a single chart. Thus, the pull_back() operation is
+   * only valid with the additional information of the chart, given by a cell
+   * on the coarse grid. An alternative implementation could shift the index
+   * depending on the coarse cell for a 1-to-1 relation between the chart space
+   * and the image space.
+   */
+  Point<dim>
+  pull_back(const typename Triangulation<dim,spacedim>::cell_iterator &cell,
+            const Point<spacedim> &p) const;
+
+  /**
+   * Push forward operation.
+   *
+   * @note This internal function is currently not compatible with the
+   * ChartManifold::pull_back() function because the given class represents an
+   * atlas of charts, not a single chart. Thus, the pull_back() operation is
+   * only valid with the additional information of the chart, given by a cell
+   * on the coarse grid. An alternative implementation could shift the index
+   * depending on the coarse cell for a 1-to-1 relation between the chart space
+   * and the image space.
+   */
+  Point<spacedim>
+  push_forward(const typename Triangulation<dim,spacedim>::cell_iterator &cell,
+               const Point<dim> &chart_point) const;
+
+  /**
+   * Gradient of the push_forward method.
+   */
+  DerivativeForm<1,dim,spacedim>
+  push_forward_gradient(const typename Triangulation<dim,spacedim>::cell_iterator &cell,
+                        const Point<dim> &chart_point) const;
+
+  /**
+   * The underlying triangulation.
+   */
+  const Triangulation<dim,spacedim> *triangulation;
+
+  /**
+   * The level of the mesh cells where the transfinite approximation is
+   * applied, usually level 0.
+   */
+  int level_coarse;
+
+  /**
+   * In case there all surrounding manifolds are the transfinite manifold or
+   * have default (invalid) manifold id, the manifold degenerates to a flat
+   * manifold and we can choose cheaper algorithms for the push_forward method.
+   */
+  std::vector<bool> coarse_cell_is_flat;
+
+  /**
+   * A MappingQ1 object that transforms between the unit and real cell for
+   * idenfifying the correct coarse cell to a given set of surrounding points.
+   */
+  MappingQGeneric<dim,spacedim> mapping_q1;
```

Comment:
Do we need this one? Wouldn't it be enough to use `StaticMappingQ1<dim,spacedim>::mapping`?